File: packages_clj/clarity/src/rlc/clarity/services/physical_object_model_service.clj
(ns rlc.clarity.services.physical-object-model-service
  (:require [clojure.tools.logging :as log]
            [clojure.pprint :refer [pprint]]
            [clojure.core.async :refer [go <!]]
            [rlc.clarity.services.entity-model-service :as e-ms]))

;; ------------------------------------------------------------------ HELPERS --

;; --------------------------------------------------------------------- KIND --

(defn retrieve-kind-of-physical-object-model
  "Retrieve and transform a physical object entity to its semantic model representation"
  [uid]
  (go
    (try
      (let [base-model (<! (e-ms/retrieve-kind-of-entity-model uid))
            ;; definitive-aspects (<! (po/get-definitive-aspects uid))
            ;; possible-aspects (<! (po/get-possible-aspects uid))
            ;; required-aspects (<! (po/get-required-aspects uid))
            ;; definitive-involvements (<! (po/get-definitive-involvements uid))
            ;; possible-involvements (<! (po/get-possible-involvements uid))
            ;; required-involvements (<! (po/get-required-involvements uid))
            ;; definitive-roles (<! (po/get-definitive-roles uid))
            ;; possible-roles (<! (po/get-possible-roles uid))
            ;; required-roles (<! (po/get-required-roles uid))
            ]
        (merge base-model
               {:category "physical object"}
               ;; {:definitive-kinds-of-aspects definitive-aspects
               ;;  :possible-kinds-of-aspects possible-aspects
               ;;  :required-kinds-of-aspects required-aspects
               ;;  :definitive-kinds-of-involvements definitive-involvements
               ;;  :possible-kinds-of-involvements possible-involvements
               ;;  :required-kinds-of-involvements required-involvements
               ;;  :definitive-kinds-of-roles definitive-roles
               ;;  :possible-kinds-of-roles possible-roles
               ;;  :required-kinds-of-roles required-roles}
               ))
      (catch Exception e))))

;; --------------------------------------------------------------- IDNIVIDUAL --

(defn retrieve-individual-physical-object-model
  "Retrieve and transform a physical object entity to its semantic model representation"
  [uid]
  (go
    (log/info "Retrieving individual physical object model" uid)
    (try
      (let [base-model (<! (e-ms/retrieve-individual-entity-model uid))
            ;; aspects (<! (po/get-definitive-aspects uid))
            ;; involvements (<! (po/get-definitive-involvements uid))
            ]
        (merge base-model
               {:category "physical object"}
               ;; {:aspects aspects
               ;;  :involvements involvements
               ))
      (catch Exception e))))

File: packages/clarity/src/services/physical-object-model.service.ts
import { Injectable, Logger } from '@nestjs/common';
import { ArchivistService } from '../archivist/archivist.service';
import { EntityModelService } from './entity-model.service';

export interface PhysicalObjectModel {
  uid: number;
  name: string;
  nature: 'kind' | 'individual';
  category: 'physical object';
  definitions?: string[];
  supertypes?: number[];
  synonyms?: string[];
  classifiers?: number[];
  aspects?: Array<{
    uid: number;
    name: string;
    relation_uid: number;
  }>;
  roles?: Array<{
    uid: number;
    name: string;
    relation_uid: number;
  }>;
  components?: Array<{
    uid: number;
    name: string;
    relation_uid: number;
  }>;
  connections?: Array<{
    uid: number;
    name: string;
    relation_uid: number;
  }>;
}

@Injectable()
export class PhysicalObjectModelService {
  private readonly logger = new Logger(PhysicalObjectModelService.name);

  constructor(
    private readonly archivistService: ArchivistService,
    private readonly entityModelService: EntityModelService,
  ) {}

  /**
   * Retrieve and transform a kind of physical object to its semantic model representation
   */
  async retrieveKindOfPhysicalObjectModel(
    uid: number,
  ): Promise<PhysicalObjectModel> {
    try {
      // Get base entity model
      const baseModel =
        await this.entityModelService.retrieveKindOfEntityModel(uid);

      // Get all facts related to this physical object
      const facts = await this.archivistService.retrieveAllFacts(uid);

      // Get aspects (things this physical object has)
      const aspects = facts
        .filter((f) => f.rel_type_name?.includes('has aspect'))
        .map((f) => ({
          uid: f.rh_object_uid,
          name: f.rh_object_name,
          relation_uid: f.rel_type_uid,
        }));

      // Get roles (roles this physical object can play)
      const roles = facts
        .filter((f) => f.rel_type_name?.includes('plays role'))
        .map((f) => ({
          uid: f.rh_object_uid,
          name: f.rh_object_name,
          relation_uid: f.rel_type_uid,
        }));

      // Get parts/components (things this object is composed of)
      const components = facts
        .filter((f) => f.rel_type_name?.includes('part of'))
        .map((f) => ({
          uid: f.lh_object_uid,
          name: f.lh_object_name,
          relation_uid: f.rel_type_uid,
        }));

      // Get connections (other objects this is connected to)
      const connections = facts
        .filter((f) => f.rel_type_name?.includes('connected'))
        .map((f) => ({
          uid: f.rh_object_uid,
          name: f.rh_object_name,
          relation_uid: f.rel_type_uid,
        }));

      this.logger.log(
        `Retrieved kind physical object model for UID ${uid}: ${baseModel.name}`,
      );

      return {
        ...baseModel,
        category: 'physical object',
        aspects,
        roles,
        components,
        connections,
      };
    } catch (error) {
      this.logger.error(
        `Failed to retrieve kind physical object model for UID ${uid}:`,
        error,
      );
      throw new Error(
        `Error retrieving kind physical object model: ${error.message}`,
      );
    }
  }

  /**
   * Retrieve and transform an individual physical object to its semantic model representation
   */
  async retrieveIndividualPhysicalObjectModel(
    uid: number,
  ): Promise<PhysicalObjectModel> {
    try {
      // Get base entity model
      const baseModel =
        await this.entityModelService.retrieveIndividualEntityModel(uid);

      // Get all facts related to this physical object
      const facts = await this.archivistService.retrieveAllFacts(uid);

      // For individuals, we might have actual aspect values rather than potential aspects
      const aspects = facts
        .filter((f) => f.rel_type_name?.includes('has aspect'))
        .map((f) => ({
          uid: f.rh_object_uid,
          name: f.rh_object_name,
          relation_uid: f.rel_type_uid,
        }));

      // Individual roles currently being played
      const roles = facts
        .filter((f) => f.rel_type_name?.includes('plays role'))
        .map((f) => ({
          uid: f.rh_object_uid,
          name: f.rh_object_name,
          relation_uid: f.rel_type_uid,
        }));

      // Actual components of this individual
      const components = facts
        .filter((f) => f.rel_type_name?.includes('part of'))
        .map((f) => ({
          uid: f.lh_object_uid,
          name: f.lh_object_name,
          relation_uid: f.rel_type_uid,
        }));

      // Actual connections of this individual
      const connections = facts
        .filter((f) => f.rel_type_name?.includes('connected'))
        .map((f) => ({
          uid: f.rh_object_uid,
          name: f.rh_object_name,
          relation_uid: f.rel_type_uid,
        }));

      this.logger.log(
        `Retrieved individual physical object model for UID ${uid}: ${baseModel.name}`,
      );

      return {
        ...baseModel,
        category: 'physical object',
        aspects,
        roles,
        components,
        connections,
      };
    } catch (error) {
      this.logger.error(
        `Failed to retrieve individual physical object model for UID ${uid}:`,
        error,
      );
      throw new Error(
        `Error retrieving individual physical object model: ${error.message}`,
      );
    }
  }
}
